File: sdk/java/src/main/java/dev/udb/client/UdbCredentials.java
package dev.udb.client;

/**
 * Mutable, thread-safe holder for the per-call auth credentials (bearer token
 * and API key). A single instance is shared across all of a {@link UdbProject}'s
 * stubs through the dynamic header interceptor built by
 * {@link UdbClient#credentialInterceptor(UdbMetadata, UdbCredentials)}, so a
 * refreshed token reaches every outbound call without rebuilding channels or
 * stubs (mirrors the TypeScript {@code core.setCredentials} hot-swap).
 */
public final class UdbCredentials {
  private volatile String bearerToken;
  private volatile String apiKey;

  public UdbCredentials(String bearerToken, String apiKey) {
    this.bearerToken = bearerToken == null ? "" : bearerToken;
    this.apiKey = apiKey == null ? "" : apiKey;
  }

  /** Seed a holder from a metadata record's static credentials. */
  static UdbCredentials fromMetadata(UdbMetadata metadata) {
    return new UdbCredentials(metadata.bearerToken(), metadata.apiKey());
  }

  public String bearerToken() {
    return bearerToken;
  }

  public String apiKey() {
    return apiKey;
  }

  /** Replace the bearer token, keeping the API key. */
  public void setBearerToken(String bearerToken) {
    this.bearerToken = bearerToken == null ? "" : bearerToken;
  }

  /** Replace the API key, keeping the bearer token. */
  public void setApiKey(String apiKey) {
    this.apiKey = apiKey == null ? "" : apiKey;
  }

  /** Replace both bearer token and API key. */
  public void set(String bearerToken, String apiKey) {
    setBearerToken(bearerToken);
    setApiKey(apiKey);
  }
}

File: sdk/csharp/Udb.Client/UdbCredentials.cs
namespace Udb.Client;

/// <summary>
/// Mutable, thread-safe holder for the per-call auth credentials (bearer token
/// and API key). A single instance is shared across a <see cref="UdbProject"/>'s
/// clients (data, auth, control-plane, storage, asset, WebRTC), so a refreshed
/// token reaches every outbound call without rebuilding channels or clients —
/// the C# analogue of the TypeScript <c>core.setCredentials</c> hot-swap.
/// </summary>
public sealed class UdbCredentials
{
    private volatile string _bearerToken;
    private volatile string _apiKey;

    public UdbCredentials(string bearerToken = "", string apiKey = "")
    {
        _bearerToken = bearerToken ?? "";
        _apiKey = apiKey ?? "";
    }

    /// <summary>Current bearer token (empty when unset).</summary>
    public string BearerToken => _bearerToken;

    /// <summary>Current API key (empty when unset).</summary>
    public string ApiKey => _apiKey;

    /// <summary>Replace both bearer token and API key.</summary>
    public void Set(string? bearerToken, string? apiKey)
    {
        _bearerToken = bearerToken ?? "";
        _apiKey = apiKey ?? "";
    }

    /// <summary>Replace just the bearer token (keeps the API key).</summary>
    public void SetBearerToken(string? bearerToken) => _bearerToken = bearerToken ?? "";

    /// <summary>Replace just the API key (keeps the bearer token).</summary>
    public void SetApiKey(string? apiKey) => _apiKey = apiKey ?? "";
}
